src/serialization/sb3.js

```diff
@@ -25,16 +27,157 @@ const {deserializeCostume, deserializeSound} = require('./deserialize-assets.js'
+ *
+ * @param {object} inputs The inputs to serialize
+ * @return {object} An object representing the serialized inputs
+ */
+const serializeInputs = function (inputs) {
+    const obj = Object.create(null);
+    for (const inputName in inputs) {
+        if (!inputs.hasOwnProperty(inputName)) continue;
+        // if block and shadow refer to the same block, only serialize one
+        if (inputs[inputName].block === inputs[inputName].shadow) {
+            // has block and shadow, and they are the same
+            obj[inputName] = [
+                INPUT_SAME_BLOCK_SHADOW,
+                inputs[inputName].block
+            ];
+        } else if (inputs[inputName].shadow === null) {
+            // does not have shadow
+            obj[inputName] = [
+                INPUT_BLOCK_NO_SHADOW,
+                inputs[inputName].block
+            ];
+        } else {
+            // block and shadow are both present and are different
+            obj[inputName] = [
+                INPUT_DIFF_BLOCK_SHADOW,
+                inputs[inputName].block,
+                inputs[inputName].shadow
+            ];
+        }
+    }
+    return obj;
+};
+
+/**
+ * Serialize the fields of a block in a more compact form.
+ * @param {object} fields The fields object to serialize
+ * @return {object} An object representing the serialized fields
+ */
+const serializeFields = function (fields) {
+    const obj = Object.create(null);
+    for (const fieldName in fields) {
+        if (!fields.hasOwnProperty(fieldName)) continue;
+        obj[fieldName] = [fields[fieldName].value];
+        if (fields[fieldName].hasOwnProperty('id')) {
+            obj[fieldName].push(fields[fieldName].id);
+        }
+    }
+    return obj;
+};
+
+/**
+ * Serialize the given block in the SB3 format with some compression of inputs,
+ * fields, and primitives.
+ * @param {object} block The block to serialize
+ * @return {object | array} A serialized representation of the block. This is an
+ * array if the block is one of the primitive types described above or an object,
+ * if not.
+ */
 const serializeBlock = function (block) {
+    const serializedPrimitive = serializePrimitiveBlock(block);
+    if (serializedPrimitive) return serializedPrimitive;
+    // If serializedPrimitive is null, proceed with serializing a non-primitive block
     const obj = Object.create(null);
-    obj.id = block.id;
     obj.opcode = block.opcode;
+    // NOTE: this is extremely important to serialize even if null;
+    // not serializing `next: null` results in strange behavior
     obj.next = block.next;
     obj.parent = block.parent;
-    obj.inputs = block.inputs;
-    obj.fields = block.fields;
+    obj.inputs = serializeInputs(block.inputs);
+    obj.fields = serializeFields(block.fields);
     obj.topLevel = block.topLevel ? block.topLevel : false;
-    obj.shadow = block.shadow;
+    obj.shadow = block.shadow; // I think we don't need this either..
```

Comment:
Can we remove it? Otherwise it's gonna be here forever (see scratchr2 with comments posed as questions from like 5 years ago and now they're still there like creepy ghosts).  Also, curious — what's the other thing the "either" is referring to?

Reply:
Oops, forgot about that comment. We do need it, so I'm just removing the comment. =]